crates/circuit/src/dag_circuit.rs

```diff
@@ -6974,6 +6861,271 @@ impl DAGCircuit {
+    last_vars: Vec<Option<NodeIndex>>,
+}
+
+impl<'a> DAGCircuitConcat<'a> {
+    /// Builds a new instance of [DAGCircuitConcat] which allows instructions to
+    /// be added continuously into the [DAGCircuit].
+    pub fn new(dag: &'a mut DAGCircuit) -> DAGCircuitConcat<'a> {
+        let num_qubits = dag.num_qubits();
+        let num_clbits = dag.num_clbits();
+        let num_vars = dag.num_vars();
+        Self {
+            dag,
+            last_qubits: (0..num_qubits).map(|_| None).collect(),
+            last_clbits: (0..num_clbits).map(|_| None).collect(),
+            last_vars: (0..num_vars).map(|_| None).collect(),
+        }
+    }
+
+    /// Finishes up the changes by re-connecting all of the ouput nodes back to the last
+    /// recorded nodes.
+    pub fn end(self) {
+        // Re-connects all of the output nodes with their respective last nodes.
+        // Add the output_nodes back to qargs
+        for (qubit, node) in self
+            .last_qubits
+            .into_iter()
+            .enumerate()
+            .filter_map(|(qubit, node)| node.map(|node| (qubit, node)))
+        {
+            let output_node = self.dag.qubit_io_map[qubit][1];
+            self.dag
+                .dag
+                .add_edge(node, output_node, Wire::Qubit(Qubit(qubit as u32)));
+        }
+
+        // Add the output_nodes back to cargs
+        for (clbit, node) in self
+            .last_clbits
+            .into_iter()
+            .enumerate()
+            .filter_map(|(clbit, node)| node.map(|node| (clbit, node)))
+        {
+            let output_node = self.dag.clbit_io_map[clbit][1];
+            self.dag
+                .dag
+                .add_edge(node, output_node, Wire::Clbit(Clbit(clbit as u32)));
+        }
+
+        // Add the output_nodes back to vars
+        for (var, node) in self
+            .last_vars
+            .into_iter()
+            .enumerate()
+            .filter_map(|(var, node)| node.map(|node| (var, node)))
+        {
+            let output_node = self.dag.var_io_map[var][1];
+            self.dag
+                .dag
+                .add_edge(node, output_node, Wire::Var(Var(var as u32)));
+        }
+    }
+
+    /// Applies a new operation to the back of the circuit. This variant works with non-owned bit indices.
+    pub fn apply_operation_back(
+        &mut self,
+        py: Python,
+        op: PackedOperation,
+        qubits: Option<OwnedOrSlice<Qubit>>,
+        clbits: Option<OwnedOrSlice<Clbit>>,
+        params: Option<SmallVec<[Param; 3]>>,
+        extra_attrs: ExtraInstructionAttributes,
+        #[cfg(feature = "cache_pygates")] py_op: Option<PyObject>,
+    ) -> NodeIndex {
+        let instruction = self.pack_instruction(
+            op,
+            qubits,
+            clbits,
+            params,
+            extra_attrs,
+            #[cfg(feature = "cache_pygates")]
+            py_op,
+        );
+        self.push_operation_back(py, instruction)
+    }
```

Comment:
I'm not sure this offers much over making `pack_instruction` public and only exposing `push_operation_back` - the output of `pack_instruction` is already effectively public because you can retrieve the `PackedInstruction` by indexing the DAG, so I'm not sure we're gaining anything by making it private.

Reply:
Hmm... sounds good, but I did want to keep an effectively similar method to `DAGCircuit::apply_operation_back`. I think that both exposing the `pack_instruction` method and keeping the `apply_operation_back` would be a viable option.